---
title: "Parse Purity Analysis w Mutation Timer"
subtitle: "DLP Mutation Time"
author: "Rhea Malhotra"
output: html_notebook
---

```{r}
library(ggplot2)
library(tidyr)
library(VariantAnnotation)
library(tidyverse)
library(argparse)
library(tidyverse)
library(scales)
library(cowplot)
library(ggpubr)
library(stringr)
library(dtplyr)
library(parallel)
source("src/global_vars.R")
```


```{r}
##Read Rdata objects into list & filter out low quality ones

##Filter the list of mutationtimer results by the aliquot IDs in the db$sequencing_bulk_dna table for which qc_status == 'Pass'

rdata_files <- list.files("/work/shah/users/chois7/retreat/germline_flt/results",pattern = ".RData",full.names = T, recursive = TRUE)
vcf_files <-  list.files("/work/shah/users/chois7/retreat/germline_flt/results",pattern = ".vcf",full.names = T, recursive = TRUE)
purity_files <-  list.files("/work/shah/users/chois7/retreat/germline_flt/results",pattern = ".purity.txt",full.names = T, recursive = TRUE)

for (i in 1:nrow(db$sequencing_bulk_dna)){
  if(db$sequencing_bulk_dna$qc_status[i] != 'Pass'){
    id <- db$sequencing_bulk_dna$isabl_id[i]
    
    filename <- paste('/work/shah/users/chois7/retreat/germline_flt/results/',id,'.vcf', sep='')
    idx <- which(vcf_files==filename)
    vcf_files <- vcf_files[-idx]
    
    filename <- paste('/work/shah/users/chois7/retreat/germline_flt/results/',id,'.purity.txt', sep='')
    idx <- which(purity_files==filename)
    purity_files <- purity_files[-idx]
    
    filename <- paste('/work/shah/users/chois7/retreat/germline_flt/results/',id,'.RData', sep='')
    idx <- which(rdata_files==filename)
    rdata_files <- rdata_files[-idx]
  }
  }



```

```{r}
#Prepare spreadsheets for figure data
library(xlsx)

Figure1 <- createWorkbook()
Figure2 <- createWorkbook()
Figure4 <- createWorkbook()
Figure5 <- createWorkbook()
ExtendedDataFigure3 <- createWorkbook()
ExtendedDataFigure6 <- createWorkbook()
ExtendedDataFigure8 <- createWorkbook()
ExtendedDataFigure9 <- createWorkbook()


dumpfile <- "/work/shah/users/chois7/retreat/purity_parsed/results/SPECTRUM-OV-002_S1_INFRACOLIC_OMENTUM_R1.RData"
if(file.exists(dumpfile)){
	#opts_chunk$set(eval=FALSE)
	load(dumpfile)
	source("src/global_vars.R")
}
```

```{r}
# Load processed data from MutationTimeR
# Load annotated VCF files for SNVs and MNVs into a list with VariantAnnotation::VCF() objects:
p <- "/work/shah/users/chois7/retreat/germline_flt/results"
d <- dir(p, pattern=".RData", full.names=TRUE)
finalSnv <- unlist(mclapply(split(d, seq_along(d) %/% 100), lapply, function(f) { # read in batches of 100
			e <- new.env()
			load(f, envir=e)
			e$vcf
		}, mc.preschedule=FALSE), recursive=FALSE)
names(finalSnv) <- sub(".conse.+","",dir(p, pattern="*.vcf.RData", full.names=FALSE))




#Load copynumber profiles as a list of `GRanges()`:
finalBB <- list()
for(f in rdata_files){
	load(f)
	colnames(mcols(bb)) <- sub("star.1","time.star",colnames(mcols(bb)) ) 
	finalBB[[f]] <- bb
}
names(finalBB) <- sub(".conse.+","",dir(p, pattern="*.bb_granges.RData", full.names=FALSE))


```

```{r}
library(spray)
whiteList <- seq_along(finalSnv) %in% 1:2703
grayList <- !whiteList

finalGenotypesSnv <- simplify2array(mclapply(finalSnv[whiteList], getGenotype, useNA="always"))
finalGenotypesIndel <- simplify2array(mclapply(finalIndel[whiteList], getGenotype, useNA="always"))
finalGenotypes <- aperm(abind::abind(subs=finalGenotypesSnv,indels=finalGenotypesIndel, along=5), c(1,5,2,3,4))
rm(finalGenotypesSnv,finalGenotypesIndel)

f <- function(x) unlist(sapply(seq_along(x), function(i) rep(i, x[i])))
d <- t(asum(finalGenotypesP[,"subs",,], 1))
o <- order(droplevels(donor2type[sample2donor[rownames(d)]]), -d[,1]/rowSums(d))
I <- t(apply(d/rowSums(d), 1, function(x) f(mg14:::roundProp(x * 100,p=100))))
d <- t(asum(finalGenotypesP[,"indels",,], 1))
J <- t(apply(d/rowSums(d), 1, function(x) if(!any(is.nan(x))) f(mg14:::roundProp(x * 100,p=100)) else rep(NA,100)))
s <- cumsum(table(droplevels(donor2type[sample2donor[rownames(d)]][o])))
```


```{r}
## SECTION 3: QC


q1 <- sapply(finalSnv, function(vcf) mean(abs(0.5- info(vcf)$pMutCNTail) > 0.495 , na.rm=TRUE))
q5 <- sapply(finalSnv, function(vcf) mean(abs(0.5- info(vcf)$pMutCNTail) > 0.475 , na.rm=TRUE))


# Doesn't run without INDELS file
# par(mfrow=c(1,1))
# boxplot(1-q5 ~ donor2type[sample2donor[names(finalSnv)]], las=2, ylab="Fraction of data inside theoretical 95% CI")
# abline(h=0.95, lty=3)

par(mfrow=c(5,5))
par(mar=c(1,1,1,1))
for(i in seq_along(finalSnv)[1:25]){
    n <- nrow(finalSnv[[i]])
    qqnorm(qnorm(info(finalSnv[[i]])$pMutCNTail[sample(1:n, min(1e4,n))]), main=paste(substr(names(finalSnv)[i],1,8), "Q5 =", signif(q5[i],2), ", Q1 =", signif(q1[i],2)), xlim=c(-5,5), ylim=c(-5,5), pch=16)
    abline(0,1, col='red')
}
```



```{r}
#SECTION 7 -- missing average ploidy
source("/work/shah/users/malhotr/PCAWG-11/code/PCAWG-functions.R")

finalPloidy <- sapply(finalBB, averagePloidy)
names(finalPloidy) <- names(finalBB)
finalHom <- sapply(finalBB, averageHom)
names(finalHom) <- names(finalBB)


```


```{r}
# SECTION 8

aggregatePerChromosome <- function(bb, isWgd=FALSE){
    .aggregateSegments <- function(m){
        #m <- mcols(bb)
        t <- weighted.mean(m$time, m$n.snv_mnv, na.rm=TRUE)
        n <- sum(m$n.snv_mnv[!is.na(m$time)], na.rm=TRUE)
        sd <- sd(m$time, na.rm=TRUE)
        ci <- weighted.mean(m$time.up-m$time.lo, m$n.snv_mnv, na.rm=TRUE)
        w <- sum(m$width[!is.na(m$time)], na.rm=TRUE)
        c(time=t, n=n, sd=sd, ci=ci,w=w)
    }
#   if(!isWgd){
    s <- split(as.data.frame(bb)[,c("time","time.up","time.lo","n.snv_mnv","width")], seqnames(bb))
    r <- t(sapply(s, .aggregateSegments))
    r <- r[c(1:22,"X"),]
#   }else{
    w <- .aggregateSegments(as.data.frame(bb))
    r <- rbind(r,WGD=w)
#   }
    return(r)
}
```

```{r}
# SECTION 8
# Aggregated timing per chromosome -- DO I NEED THIS SECTION

allChrAgg <- simplify2array(mclapply(finalBB, aggregatePerChromosome, mc.cores=2))

t <- allChrAgg[1:23,"time",!isWgd]
t[allChrAgg[1:23,"w",!isWgd] < diff(chrOffset)[1:23]*.33] <- NA

s <- split(as.data.frame(t(t)), droplevels(donor2type[sample2donor[names(finalSnv)]])[!isWgd])
n <- 10


at <- function(x, n){
    if(sum(!is.na(x))<3) return(rep(sum(!is.na(x))/n,n))
    bw=if(sum(!is.na(x))< 6) 0.5 else "nrd0"
    d <- density(x, n=n, from=1/n/2, to=1-1/n/2, bw=bw, na.rm=TRUE)
    d$y/sum(d$y)*d$n
}

allChrCancerHist <- sapply(s, apply, 2, at, n=n, simplify="array")
u <- split(data.frame(WGD=allChrAgg["WGD","time",isWgd]), droplevels(donor2type[sample2donor[names(finalSnv)]])[isWgd])
wgdCancerHist <- sapply(u, function(x) if(nrow(x)>0){at(x$WGD,n=n)}else{rep(0,n)}, simplify="array")
allChrCancerHist <- abind::abind(allChrCancerHist, All=sapply(sapply(s, as.matrix), at, n=n, simplify="array")/23*5, WGD=wgdCancerHist, along=2)
```
```{r}
# SECTION 8
# per tumor type


prgn <- RColorBrewer::brewer.pal(11,"PRGn")
set1 <- RColorBrewer::brewer.pal(9,"Set1")
col <- colorRampPalette(set1[c(4,9,3)])(n)

p <- 0
v <- table(droplevels(donor2type[sample2donor[names(finalSnv)]]))
h <- (allChrCancerHist + p)  / rep(v + p, each=prod(dim(allChrCancerHist)[1:2]))
h <- aperm(h, c(2,3,1))

a <- colMeans(h[c("All","WGD"),,] * c(23/5,1)) %*% 1:n / asum(h* c(23/5,1), c(1,3))
o <- order(-a)
h <- h[,o,]
w <- v[o]>=15 & apply(h, 2, max) > 0.05*8/n
h <- h[,w,]

m <- 0.02
layout(matrix(1:prod(dim(h)[1:2]+1), ncol=dim(h)[1]+1, byrow=TRUE), height=c(rev(apply(h, 2, max))+m, 0.15), width=c(5, rep(1,dim(h)[1])))
par(mar=c(0.05,0.1,0,0.1), xpd=NA)
for(j in dim(h)[2]:0+1) for(i in 0:dim(h)[1]+1) {
        #if(all(h[i,j,]==0)) 
        if(i==1 & j !=1) {plot(NA,NA,xlab="",ylab="", xaxt="n",yaxt="n",xlim=c(0,1),ylim=c(0,1), bty="n")
            text(1,0,dimnames(h)[[2]][j-1],pos=2)
            next
        }
        if(j ==1 ){
            plot(NA,NA,xlab="",ylab="", xaxt="n",yaxt="n",xlim=c(0,1),ylim=c(0,1), bty="n")
            if(i==1) next
            text(0.5,1,dimnames(h)[[1]][i-1],pos=1)
            next
        }
        r <- c(0,max(h[,j-1,]+m))
        par(bty=if(i==2)"L" else "n")
        barplot(h[i-1,j-1,], ylim=r, width=1/n,space=0, col=rev(col), xaxt="n", yaxt="n", xlab="",ylab="", border=NA,xpd=TRUE, yaxs="i", xaxs="i", xlim=c(-0.5/n,1+0.5/n))
        axis(side=1, at=c(-0.5/n,1+0.5/n), labels=c("",""), tcl=-.1)
        if(i>1)
            abline(v=0, col='lightgrey', lty=3)
        if(i==2){
            abline(h=0.05*8/n, col='lightgrey', lty=1)
            axis(side=2, at=c(0,0.05*8/n), labels=c("",""), tcl=-.1)
        }
    }
```
```{r}
vv <- v[dimnames(h)[[2]]]
vv <- vv/sum(vv)

hh <- matrix(matrix(aperm(h, c(1,3,2)), ncol=length(vv)) %*% vv, nrow=nrow(h))
rownames(hh) <- rownames(h)

par(mar=c(3,3,1,1), mgp=c(2,.5,0), tcl=-0.5, bty="L", xpd=NA)
barplot(hh["WGD",], space=0, col=rev(col), xlab="Time [mutations]", ylab="Relative frequency", width=0.1, ylim=c(0,.065), yaxs='r', border=NA)
axis(side=1)
```




